bluetooth: host: Allow for ECDH operations through system workq

MarekPieta opened the pull request:
Change allows performing ECDH operations through system workq. This is done to allow reducing memory consumption by disabling the long workq on small SoCs.

MarekPieta committed:
bluetooth: host: Allow for ECDH operations through system workq

Change allows performing ECDH operations through system workq. This is
done to allow reducing memory consumption by disabling the long workq on
small SoCs.

Signed-off-by: Marek Pieta <Marek.Pieta@nordicsemi.no>

jhedberg reviewed (dismissed):
I'll approve, but seems like we've moved quite far from recommending to treat the system wq as an ISR, i.e. something where you shouldn't block for a long time, let alone perform operations where you don't even yield for a long time.

Thalley commented:
> I'll approve, but seems like we've moved quite far from recommending to treat the system wq as an ISR, i.e. something where you shouldn't block for a long time, let alone perform operations where you don't even yield for a long time.

Same reason why I didn't approve. 

Perhaps we need to at least enable some tests running this configuration to ensure that it at least kind of works?

MarekPieta commented:
> I'll approve, but seems like we've moved quite far from recommending to treat the system wq as an ISR, i.e. something where you shouldn't block for a long time, let alone perform operations where you don't even yield for a long time.

By default we still use the long workqueue for ECC crypto. The PR allows optimizing RAM usage on smaller/simpler targets and application developer enables it on own risk.

MarekPieta commented:
> > I'll approve, but seems like we've moved quite far from recommending to treat the system wq as an ISR, i.e. something where you shouldn't block for a long time, let alone perform operations where you don't even yield for a long time.
> 
> Same reason why I didn't approve.
> 
> Perhaps we need to at least enable some tests running this configuration to ensure that it at least kind of works?

Maybe we should enable it e.g. in some of BabbleSim tests?

jhedberg commented:
> Maybe we should enable it e.g. in some of BabbleSim tests?

Yes, please

MarekPieta commented:
> > Maybe we should enable it e.g. in some of BabbleSim tests?
> 
> Yes, please

Could you advise me what test should be used to verify the configuration?

jhedberg commented:
> > > Maybe we should enable it e.g. in some of BabbleSim tests?
> > 
> > 
> > Yes, please
> 
> Could you advise me which BabbleSim test should be used to verify the configuration, please?

I'm not really that familiar with the bsim test cases, but just looking at what's there, maybe something like `tests/bsim/bluetooth/host/gatt/notify` and `tests/bsim/bluetooth/host/security/ccc_update` would be ok?

Thalley commented:
> > > > Maybe we should enable it e.g. in some of BabbleSim tests?
> > > 
> > > 
> > > Yes, please
> > 
> > 
> > Could you advise me which BabbleSim test should be used to verify the configuration, please?
> 
> I'm not really that familiar with the bsim test cases, but just looking at what's there, maybe something like `tests/bsim/bluetooth/host/gatt/notify` and `tests/bsim/bluetooth/host/security/ccc_update` would be ok?

Any tests that does encryption should be OK :)

MarekPieta committed:
tests: bsim: bluetooth: Test with CONFIG_BT_LONG_WQ disabled

Change adds bsim tests with CONFIG_BT_LONG_WQ=n to ensure correct
behavior. Change uses host/security/ccc_update test.

Change also switches to using more unique simulation ID in the test
scripts.

Signed-off-by: Marek Pieta <Marek.Pieta@nordicsemi.no>

MarekPieta committed:
tests: bsim: bluetooth: Test with CONFIG_BT_LONG_WQ disabled

Change adds bsim tests with CONFIG_BT_LONG_WQ=n to ensure correct
behavior. Change uses host/gatt/ccc_store test.

Change also switches to using more unique simulation ID in the test
scripts.

Signed-off-by: Marek Pieta <Marek.Pieta@nordicsemi.no>

MarekPieta commented:
Disabled the `CONFIG_BT_LONG_WQ` for 2 bsim tests. Please take a look

jhedberg reviewed (dismissed):
Looks fine to me, assuming the bsim CI tests will pass

MarekPieta commented on tests/bsim/bluetooth/host/security/ccc_update/test_scripts/ccc_update_no_long_wq.sh:
```diff
@@ -0,0 +1,53 @@
+
+test_name='ccc_update'
+test_exe="bs_${BOARD_TS}_tests_bsim_bluetooth_host_security_${test_name}_overlay-no_long_wq_conf"
+simulation_id="${test_name}_3"
```
Applied the renames for both modified tests

Thalley commented on tests/bsim/bluetooth/host/gatt/ccc_store/testcase.yaml:
```diff
@@ -15,3 +15,8 @@ tests:
+  bluetooth.host.gatt.ccc_store_3:
+    harness_config:
+      bsim_exe_name: tests_bsim_bluetooth_host_gatt_ccc_store_overlay-no_long_wq_conf
+    extra_args:
+      EXTRA_CONF_FILE=overlay-no_long_wq.conf
```
Just FYI, you could also have done the following (I think)

```suggestion
    extra_args:
      - CONFIG_BT_LONG_WQ=n
```

If you didn't want that extra config file

MarekPieta replied:
I am aware that we can use `extra_configs` here. I aligned my PR with the existing approach (there are already Kconfig overlay files that modify value of a single Kconfig option and introduce a comment).

MarekPieta commented:
Applied changes requested by @aescolar. Please rereview

aescolar commented on tests/bsim/bluetooth/host/security/ccc_update/test_scripts/ccc_update_no_long_wq.sh:
```diff
@@ -0,0 +1,53 @@
+Execute ./bs_2G4_phy_v1 -v=${verbosity_level} -s=${simulation_id} \
+  -D=3 -sim_length=60e6
+
+if [ "${1}" == 'debug0' ]; then
```
Fair enough, note I was not requesting a change. I'm ok with this merging kind of things, they just seem a bit funny.
